packages/enclave-contracts/contracts/registry/BondingRegistry.sol

```diff
@@ -0,0 +1,757 @@
+    }
+
+    /// @notice Get operator's ticket balance at a specific block
+    /// @dev Uses checkpoint mechanism from ticket token
+    /// @param operator Address of the operator
+    /// @param blockNumber Block number to query
+    /// @return Ticket balance at the specified block
+    function getTicketBalanceAtBlock(
+        address operator,
+        uint256 blockNumber
+    ) external view returns (uint256) {
+        return ticketToken.getPastVotes(operator, blockNumber);
+    }
+
+    /// @notice Get operator's total pending exit amounts
+    /// @param operator Address of the operator
+    /// @return ticket Total pending ticket balance in exit queue
+    /// @return license Total pending license bond in exit queue
+    function pendingExits(
+        address operator
+    ) external view returns (uint256 ticket, uint256 license) {
+        return _exits.getPendingAmounts(operator);
+    }
+
+    /// @notice Preview how much an operator can currently claim
+    /// @param operator Address of the operator
+    /// @return ticket Claimable ticket balance
+    /// @return license Claimable license bond
+    function previewClaimable(
+        address operator
+    ) external view returns (uint256 ticket, uint256 license) {
+        return _exits.previewClaimableAmounts(operator);
+    }
+
+    /// @inheritdoc IBondingRegistry
+    function isLicensed(address operator) external view returns (bool) {
+        return operators[operator].licenseBond >= _minLicenseBond();
+    }
+
+    /// @inheritdoc IBondingRegistry
+    function isRegistered(address operator) external view returns (bool) {
+        return operators[operator].registered;
+    }
+
+    /// @inheritdoc IBondingRegistry
+    function isActive(address operator) external view returns (bool) {
+        return operators[operator].active;
+    }
+
+    /// @inheritdoc IBondingRegistry
+    function hasExitInProgress(address operator) external view returns (bool) {
+        Operator storage op = operators[operator];
+        return op.exitRequested && block.timestamp < op.exitUnlocksAt;
+    }
+
+    // ======================
+    // Operator Functions
+    // ======================
+
+    /// @inheritdoc IBondingRegistry
+    function registerOperator() external noExitInProgress(msg.sender) {
+        // Clear previous exit request
+        if (operators[msg.sender].exitRequested) {
+            operators[msg.sender].exitRequested = false;
+            operators[msg.sender].exitUnlocksAt = 0;
+        }
+
+        require(
+            !ISlashingManager(slashingManager).isBanned(msg.sender),
+            CiphernodeBanned()
+        );
+        require(!operators[msg.sender].registered, AlreadyRegistered());
+        require(
+            operators[msg.sender].licenseBond >= licenseRequiredBond,
+            NotLicensed()
+        );
+
+        operators[msg.sender].registered = true;
+
+        if (address(registry) != address(0)) {
```

Comment:
perhaps we should add the check inside setRegistry instead? so it can never be zero and we avoid the check here for every call by ciphernodes? or are there cases where we wouldn't have a ciphernode registry set?